### PR TITLE
mul/ewise rules for basic arithmetic semiring

### PR DESCRIPTION
**Pull request by Wimmerer**
I removed some I'm still testing to get feedback on these and avoid a monster PR.

Notes:
1. You need to call `test_*rule` with `check_inferred=false`. Issue #25 will fix.
2. Missing the kwargs. I want to get everything working first before trying those.
3. rrules are incorrect according to tests. Some of these are just floating point issues. However for `mul` there's a deeper issue. I'm 85-90% sure the rules are correct, but the patterns are not the same as for FiniteDifferences, and occasionally there's different values.

### CODE REVIEW

**Comment by Wimmerer** [comment]
I apologize for the messy PR, the only important parts are in the tests and chainrules folders. 

I'm primarily interested in your thoughts about the `rrules` for `mul`, and in particular whether I'm wrong, FiniteDifferences is wrong, or I just haven't given FiniteDifferences the right information. 

Everything works fine for dense. For sparse inputs though there's two problems:
1.  ∂A has a different sparsity pattern, and thus different values where the sparsity is different.
2.  ∂B is straight up wrong according to FiniteDifferences. 

@mzgubic

File: test/runtests.jl [hunk 1]
```diff
@@ -14,4 +14,5 @@ println("Testing SuiteSparseGraphBLAS.jl")
 @testset "SuiteSparseGraphBLAS" begin
     include_test("gbarray.jl")
     include_test("operations.jl")
+    include_test("testrules.jl")
```
Comment (mzgubic) [inline test/runtests.jl:17]: Usually the structure of `test` folder mirrors the `src/` folder, which makes it easier to find things when the package grows.

File: src/chainrules/ewiserules.jl [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+end
+
+function rrule(::typeof(eadd), A::GBArray, B::GBArray, ::typeof(BinaryOps.PLUS))
+    function pluspullback(ΔΩ)
+        return NoTangent(), ΔΩ, ΔΩ, NoTangent()
+    end
+    return eadd(A, B, BinaryOps.PLUS), pluspullback
+end
+
+# Do I have to duplicate this? I get 4 tangents instead of 3 if I call the previous rule.
```
Comment (mzgubic) [inline src/chainrules/ewiserules.jl:58]: Yes, thought not entirely - we can call the previous rule and just take the first 3 gradients

File: src/chainrules/ewiserules.jl [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+function rrule(::typeof(eadd), A::GBArray, B::GBArray)
+    function pluspullback(ΔΩ)
+        return NoTangent(), ΔΩ, ΔΩ
+    end
```
Comment (mzgubic) [inline src/chainrules/ewiserules.jl:62]: Are these correct? My impression is they should be something like
```julia
    function pluspullback(ΔΩ)
        IA, JA, XA = findnz(A)
        dA = GBMatrix(IA, JA, ones(size(XA)))
        IB, JB, XB = findnz(B)
        dB = GBMatrix(IB, JB, ones(size(XB)))
        return NoTangent(), dA, dB
    end
```

**Review by mzgubic — COMMENTED** [review]
Didn't manage to finish it all today, but here's a few comments

File: Project.toml [hunk 4]
```diff
@@ -5,7 +5,11 @@ version = "0.4.0"
 
 [deps]
 CEnum = "fa961155-64e5-5f13-b03f-caf6b980ea82"
+ChainRules = "082447d4-558c-5d27-93f4-14fc19e9eca2"
```
[the remark this answers is not in the document]
Comment (Wimmerer) [reply Project.toml:8]: No we don't. I thought I had that in my upper level environment not this one.

File: src/chainrules/chainruleutils.jl [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+const RealOrComplex = Union{Real, Complex}
+
+#Required for ChainRulesTestUtils
+function FiniteDifferences.to_vec(M::GBMatrix)
+    I, J, X = findnz(M)
+    function backtomat(xvec)
+        return GBMatrix(I, J, xvec; nrows = size(M, 1), ncols = size(M, 2))
+    end
+    return X, backtomat
+end
```
[the remark this answers is not in the document]
Comment (Wimmerer) [reply src/chainrules/chainruleutils.jl:14]: I'm not sure I understand what you mean array of structs here. I'll try this though.

File: src/chainrules/ewiserules.jl (hunk 2 above)
Comment (Wimmerer) [reply src/chainrules/ewiserules.jl:58]: Do I call the pullback to obtain those gradients?

Comment (mzgubic) [reply src/chainrules/ewiserules.jl:58]: Yep, an example is here:
https://github.com/JuliaDiff/ChainRules.jl/blob/0e8e4751fef00b81eaf9ddb2baf2a0327d04251b/src/rulesets/LinearAlgebra/dense.jl#L217-L238

File: src/chainrules/chainruleutils.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (mzgubic) [reply src/chainrules/chainruleutils.jl:14]: Ah, never mind. I meant something like
```julia
julia> struct Fred
           a::Float64
       end

julia> Base.zero(::Fred) = Fred(0)

julia> m = GBMatrix([1, 2], [2, 3], [Fred(3.2), Fred(4.3)])
ERROR: ArgumentError: Not a valid GrB data type
Stacktrace:
 [1] toGBType(x::Type)
   @ SuiteSparseGraphBLAS ~/JuliaEnvs/SuiteSparseGraphBLAS.jl/src/gbtypes.jl:132
 [2] GBMatrix{Fred}(nrows::Int64, ncols::Int64)
   @ SuiteSparseGraphBLAS ~/JuliaEnvs/SuiteSparseGraphBLAS.jl/src/matrix.jl:9
 [3] GBMatrix(I::Vector{Int64}, J::Vector{Int64}, X::Vector{Fred}; dup::SuiteSparseGraphBLAS.Types.PLUS_T, nrows::Int64, ncols::Int64)
   @ SuiteSparseGraphBLAS ~/JuliaEnvs/SuiteSparseGraphBLAS.jl/src/matrix.jl:24
 [4] GBMatrix(I::Vector{Int64}, J::Vector{Int64}, X::Vector{Fred})
   @ SuiteSparseGraphBLAS ~/JuliaEnvs/SuiteSparseGraphBLAS.jl/src/matrix.jl:24
 [5] top-level scope
   @ REPL[75]:1
```
but this is not supported anyway, so fine to just keep it as it is.

File: src/chainrules/ewiserules.jl [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    return eadd(A, B, BinaryOps.PLUS), pluspullback
+end
```
Comment (mzgubic) [inline src/chainrules/ewiserules.jl:64]: What's happening here is that the `A` and `B` have some sparsity pattern, and the `y = eadd(A, B)` has a "combined" sparsity pattern.

`dA` needs to preserve the sparsity pattern of `A`.
e.g. for
```julia
julia> X = GBMatrix([1,], [1,], [1.0,]; nrows=2, ncols=2);

julia> Y = GBMatrix([2,], [2,], [1.0,]; nrows=2, ncols=2);
```

the following rule works
```julia
function rrule(::typeof(eadd), A::GBArray, B::GBArray)
    function pluspullback(ΔΩ)
        IA, JA, XA = findnz(A)
        dA = GBMatrix(IA, JA, [ΔΩ[i, j] for (i, j) in zip(IA, JA)]; nrows=2, ncols=2)
        IB, JB, XB = findnz(B)
        dB = GBMatrix(IB, JB, [ΔΩ[i, j] for (i, j) in zip(IB, JB)]; nrows=2, ncols=2)
        return NoTangent(), dA, dB
    end
    return eadd(A, B, BinaryOps.PLUS), pluspullback
end
```
This is a horrendous implementation of course, but just to illustrate the issue.

Comment (Wimmerer) [reply src/chainrules/ewiserules.jl:64]: Ah ok I just need a mask then. That's easy enough.

**Comment by mzgubic** [comment]
>Everything works fine for dense. For sparse inputs though there's two problems:
∂A has a different sparsity pattern, and thus different values where the sparsity is different.
∂B is straight up wrong according to FiniteDifferences.

I think the underlying issue is the same (and also the same one as in the elementwise rules). What it comes down is an instance of the "array dilemma", discussed in great detail over many issues and PRs. See https://github.com/JuliaDiff/ChainRulesCore.jl/pull/347 (and related issues) for a discussion, but I warn you, it is a rabbit hole ;)

Essentially what it comes down to is whether you think of the input, say `A::GBMatrix` as an efficient representation of an array that just happens to be sparse, or whether you think of it as a struct. Consider `y = A * B`, where `B::Matrix` is a dense array, and `y` is therefore dense as well.

Primal computation will be fast because `A` is sparse. In fact `A` was probably chosen to be `GBMatrix` solely to get 
that speedup. What happens in the backward pass depends on how you interpret `A`: an array, or a struct?

if you interpret it as an array, the `dA = mul(ΔΩ, B')` will be dense and you lose all the benefits of the speedup, but `dA` will match the `dA` you would have gotten with a dense `A` with the zeros in the same place as structural zeros of the sparse `A`.

If you interpret it as a struct, meaning that the zeros are structural, it doesn't make sense to compute the tangents to all the zeros, and you can compute the backward pass efficiently. Since `dA` for sparse `A` is sparse in this case, it is somewhat unintuitive that it is different to the `dA` that would be obtained if `A` was a dense array with the zeros in the same place.

Long story short, we are treating them as structs now in order to not completely kill efficiency. We should probably treat them as structs here as well.

Aside: [projection](https://github.com/JuliaDiff/ChainRulesCore.jl/pull/385), merged recently, was a way to make sure rules with abstractly typed arguments still return the correct tangent type. The classic example is `Diagonal * Matrix` where we project the dense gradient onto the `Diagonal`.

----------------

In this case, as you point out, masking is all we need to do, since we are writing dedicated rules for `GBMatrix` multiplication.

File: Project.toml [hunk 7]
```diff
@@ -15,8 +19,8 @@ SparseArrays = "2f01184e-e22b-5df5-ae63-d93ebab69eaf"
 Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
 
 [compat]
```
[the remark this answers is not in the document]
Comment (Wimmerer) [reply Project.toml:21]: I'll update these in a different PR.

File: src/chainrules/chainruleutils.jl [hunk 8]
```diff
@@ -0,0 +1,46 @@
+import FiniteDifferences
+import LinearAlgebra
+import ChainRulesCore: frule, rrule
+using ChainRulesCore
+const RealOrComplex = Union{Real, Complex}
+
+#Required for ChainRulesTestUtils
+function FiniteDifferences.to_vec(M::GBMatrix)
```
[the remark this answers is not in the document]
Comment (Wimmerer) [reply src/chainrules/chainruleutils.jl:8]: Done

**Comment by Wimmerer** [comment]
@mzgubic fixed up the rules and the test structure relevant to this PR. I added constructor rules, although they should've likely been separate PR :/.

However, they are broken for reasons I cannot figure out `UInt64(0.99)`, inexact is the error. There should never be a construction where `I` and/or `J` are float vectors, and `to_vec` and `rand_tangent` don't seem to allow this to happen anyway.

I'm trying to finish up all the other rules/semirings except possibly map by Tuesday (I'd like to have a call about map, as it effects non-AD parts of the package as well).
